core/src/main/java/org/apache/accumulo/core/singletons/SingletonManager.java

```diff
@@ -0,0 +1,190 @@
+    CLIENT,
+    /**
+     * In this mode singletons are never disabled.
+     */
+    SERVER,
+    /**
+     * In this mode singletons are never disabled unless the mode is set back to CLIENT. The user
+     * can do this by using util.CleanUp (an old API created for users).
+     */
+    CONNECTOR
+  }
+
+  private static long reservations;
+  private static Mode mode;
+  private static boolean enabled;
+  private static boolean transitionedFromClientToConnector;
+  private static List<SingletonService> services;
+
+  @VisibleForTesting
+  static void reset() {
+    reservations = 0;
+    mode = Mode.CLIENT;
+    enabled = true;
+    transitionedFromClientToConnector = false;
+    services = new ArrayList<>();
+  }
+
+  static {
+    reset();
+  }
+
+  private static void enable(SingletonService service) {
+    try {
+      service.enable();
+    } catch (RuntimeException e) {
+      // TODO log
+      e.printStackTrace();
+    }
+  }
+
+  private static void disable(SingletonService service) {
+    try {
+      service.disable();
+    } catch (RuntimeException e) {
+      // TODO log
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Register a static singleton that should be disabled and enabled as needed.
+   */
+  public static synchronized void register(SingletonService service) {
+    if (enabled && !service.isEnabled()) {
+      enable(service);
+    }
+
+    if (!enabled && service.isEnabled()) {
+      disable(service);
+    }
+
+    services.add(service);
+  }
+
+  /**
+   * This method should be called when creating Accumulo clients using the public API. Accumulo
+   * clients created internally within the code probably should not call this method. While a client
+   * holds a reservation, singleton services are enabled.
+   *
+   * @return A reservation that must be closed when the AccumuloClient is closed.
+   */
+  public static synchronized SingletonReservation getClientReservation() {
+    Preconditions.checkState(reservations >= 0);
+    reservations++;
+    transition();
+    return new SingletonReservation();
+  }
+
+  static synchronized void releaseRerservation() {
+    Preconditions.checkState(reservations > 0);
+    reservations--;
+    transition();
+  }
+
+  @VisibleForTesting
+  public static long getReservationCount() {
+    return reservations;
+  }
+
+  /**
+   * Change how singletons are managed. The default mode is {@link Mode#CLIENT}
+   */
+  public static synchronized void setMode(Mode mode) {
+    if (SingletonManager.mode == Mode.CLIENT && mode == Mode.CONNECTOR) {
+      if (transitionedFromClientToConnector) {
+        throw new IllegalStateException("Can only transition from " + Mode.CLIENT + " to "
+            + Mode.CONNECTOR + " once.  This error indicates that "
+            + "org.apache.accumulo.core.util.CleanUp.shutdownNow() was called and then later a "
+            + "Connector was created.  Connectors can not be created after CleanUp.shutdownNow()"
+            + " is called.");
+      }
```

Comment:
Allowing reservations to transition state is confusing and seems problematic.  If this transition is done for a special case, then couldn't you tear down and create a new reservation instead?  It seems like making the mode of a reservation final would simplify things.

Reply:
Reservations are separate from the mode.  Reservations track how many active Accumulo Clients there are.  Reservations only matter when the mode is client.   When the mode is connector or server, then singleton services are never disabled even if the reservation count is zero.

Reply:
Ah I see.  Thanks for clarifying.